### PR TITLE
[JEP416] Add caller sensitive adapter method in java.lang.ClassLoader

### PR DESCRIPTION
**Pull request by babsingh**
For JEP416, a private adapter method, which accepts an additional `Class<?>`
parameter, is needed for `ClassLoader.registerAsParallelCapable(Class<?>)`.

`@CallerSensitiveAdapter` is a variant of `@CallerSensitive`. So, treat
`@CallerSensitiveAdapter` similar to `@CallerSensitive` in
- `inlInternalsNewInstanceImpl`
- `inlVMGetStackClassLoader`
- `inlVMGetStackClass`

Closes: https://github.com/eclipse-openj9/openj9/issues/14133
Related: https://github.com/eclipse-openj9/openj9/issues/13852

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
fyi @JasonFengJ9 @fengxue-IS @tajila @pshipton

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 1]
```diff
@@ -396,6 +400,13 @@ public Object run() {
  */
 @CallerSensitive
 public static Class<?> forName(String className) throws ClassNotFoundException {
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/Class.java:402]: Is this change required for `forName(Module module, String name)` as well?

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 2]
```diff
@@ -396,6 +400,13 @@ public Object run() {
...
+	return forName(className, null);
+}
+
+/*[IF JAVA_SPEC_VERSION >= 18]*/
+@CallerSensitiveAdapter
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/Class.java:407]: `getStackClass()` checks the annotation `@CallerSensitive`, now `@CallerSensitiveAdapter` need to be checked instead.

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 3]
```diff
@@ -474,6 +487,15 @@ public Field run() {
...
 	throws ClassNotFoundException
+{
+	return forName(className, initializeBoolean, classLoader, null);
+}
+
+/*[IF JAVA_SPEC_VERSION >= 18]*/
+@CallerSensitiveAdapter
+/*[ENDIF] JAVA_SPEC_VERSION >= 18 */
+private static Class<?> forName(String className, boolean initializeBoolean, ClassLoader classLoader, Class<?> caller)
+	throws ClassNotFoundException
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/Class.java:498]:
```suggestion
/*[IF JAVA_SPEC_VERSION >= 18]*/
{
	return forName(className, initializeBoolean, classLoader, null);
}

@CallerSensitiveAdapter
private static Class<?> forName(String className, boolean initializeBoolean, ClassLoader classLoader, Class<?> caller)
	throws ClassNotFoundException
/*[ENDIF] JAVA_SPEC_VERSION >= 18 */
```
This keeps pre-JDK18 code intact.
Similar comments for other private method additions.

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 1 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Class.java:402]: yes; updated PR.

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 2 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Class.java:407]: updated.

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 3 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Class.java]: new changes keep pre-jdk18 code intact.

**Comment by babsingh** [comment]
I have divided this work into two PRs. This PR contains the changes needed to fix the deadlock in https://github.com/eclipse-openj9/openj9/issues/14133. The other PR is https://github.com/eclipse-openj9/openj9/pull/14267, which adds caller sensitive adapter methods in j.l.Class. I have opened #14267 since I am encountering an issue while adding caller sensitive adapter methods in j.l.Class. More details about the issue will be discussed in #14267. @JasonFengJ9 @tajila Can you please review this PR?

**Comment by JasonFengJ9** [comment]
Oops, missed this one, will look at it this afternoon.

File: runtime/bcutil/ClassFileOracle.cpp [hunk 4]
```diff
@@ -51,6 +51,11 @@ ClassFileOracle::KnownAnnotation ClassFileOracle::_knownAnnotations[] = {
 #define JDK_INTERNAL_REFLECT_CALLERSENSITIVE_SIGNATURE "Ljdk/internal/reflect/CallerSensitive;"
 		{JDK_INTERNAL_REFLECT_CALLERSENSITIVE_SIGNATURE, sizeof(JDK_INTERNAL_REFLECT_CALLERSENSITIVE_SIGNATURE)},
 #undef JDK_INTERNAL_REFLECT_CALLERSENSITIVE_SIGNATURE
+#if JAVA_SPEC_VERSION >= 18
+#define JDK_INTERNAL_REFLECT_CALLERSENSITIVEADAPTER_SIGNATURE "Ljdk/internal/reflect/CallerSensitiveAdapter;"
+		{JDK_INTERNAL_REFLECT_CALLERSENSITIVEADAPTER_SIGNATURE, sizeof(JDK_INTERNAL_REFLECT_CALLERSENSITIVEADAPTER_SIGNATURE)},
```
Comment (JasonFengJ9) [inline runtime/bcutil/ClassFileOracle.cpp:56]: I know I asked at https://github.com/eclipse-openj9/openj9/pull/14177#discussion_r771714437 to check `@CallerSensitiveAdapter`, at that time, it precedes w/ `getStackClass()`.
This PR doesn't have a caller sensitive method annotated by `@CallerSensitiveAdapter`. Are the native changes required to fix `java/lang/ClassLoader/deadlock/DelegateTest.java` hang, or just prepare for https://github.com/eclipse-openj9/openj9/pull/14267 instead?

Comment (babsingh) [reply runtime/bcutil/ClassFileOracle.cpp:56]: `ClassLoader.registerAsParallelCapable(Class<?>)` is annotated with `@CallerSensitiveAdapter`, and `ClassLoader.registerAsParallelCapable()` is annotated with `@CallerSensitive`. We also need the native changes for #14267.

Comment (JasonFengJ9) [reply runtime/bcutil/ClassFileOracle.cpp:56]: > We also need the native changes for #14267

Can the `ClassLoader` change alone address the hang failure, does the native change fix any test failure?

Comment (babsingh) [reply runtime/bcutil/ClassFileOracle.cpp:56]: Yes, it can alone address the hang failure. Without the native change, `ClassLoader.registerAsParallelCapable(Class<?>)` won't be identified as a caller sensitive method by `cInterpGetStackClassJEP176Iterator` and other dependencies. This may lead to errors elsewhere.

**Review by JasonFengJ9 — APPROVED** [review]
LGTM

**Comment by tajila** [comment]
Jenkins test sanity win jdk8

**Comment by tajila** [comment]
Jenkins test sanity plinux jdk17

**Comment by tajila** [comment]
Jenkins test sanity xlinux jdk18

**Review by tajila — APPROVED** [review]
